### PR TITLE
Add scopes and change routes for group leaderboard

### PR DESCRIPTION
**Pull request by weiqingtoh**
This PR: 
- Adds calculated attributes and scopes to sort the group for the group leaderboard.

The calculated attributes are a little nasty since we had to manually calculate the 'calculated attributes' (achievement count and experience points) for each course user, so I've documented down the reasons for some of the scopes. 

Related to #634

### CODE REVIEW

**Comment by weiqingtoh** [comment]
Actually there are quite a lot of nasty calculated attributes and scopes, should I move them into a concern instead? Might be cleaner.

File: app/models/course/group.rb [hunk 1]
```diff
@@ -15,6 +15,57 @@ class Course::Group < ActiveRecord::Base
   validate :validate_new_users_are_unique
   validate :validate_presence_of_group_manager, on: :update
 
+  # @!attribute [r] average_experience_points
+  #   Returns the average experience points of group users in this group.
+  #   Notes:
+  #     - course_user.role == 0 is used to ensure that this applies only to students only.
```
Comment (allenwq) [inline app/models/course/group.rb]: Move these comments and put them above the lines in the source code ?

File: app/models/course/group.rb [hunk 2]
```diff
@@ -15,6 +15,57 @@ class Course::Group < ActiveRecord::Base
...
+  #     - CAST is used to force a float division (integer division by default).
+  #     - greatest(#, 1) is used to avoid division by 0.
+  calculated :average_experience_points, (lambda do
+    Course::GroupUser.where { group_id == course_groups.id }.
+      joins { course_user.experience_points_records.outer }.
+      where { course_user.role == 0 }.select(<<-SQL)
```
Comment (allenwq) [inline app/models/course/group.rb]: Better to reference the role id using symbols: `CourseUser.roles[:student]`

**Comment by allenwq** [comment]
@weiqingtoh In this case I think it's still fine. we should split them when the model becomes very big and not manageable.

File: app/models/course/group.rb (hunk 1 above)
Comment (weiqingtoh) [inline app/models/course/group.rb]: I can move some of them, but a little hard since I use heredocs >_<

Comment (allenwq) [inline app/models/course/group.rb]: The thing is if your comments is far away from source code, it will be easily forget and become outdated after source code modification. and it's also not very readable since there is no context.

**Comment by allenwq** [comment]
@weiqingtoh Thanks !

File: app/models/course/group.rb [hunk 3]
```diff
@@ -15,6 +15,57 @@ class Course::Group < ActiveRecord::Base
...
 
+  # @!attribute [r] average_experience_points
+  #   Returns the average experience points of group users in this group who are students.
+  calculated :average_experience_points, (lambda do
+    Course::GroupUser.where { group_id == course_groups.id }.
+      joins { course_user.experience_points_records.outer }.
+      where { course_user.role == CourseUser.roles[:student] }.
+      # CAST is used to force a float division (integer division by default).
+      # greatest(#, 1) is used to avoid division by 0.
+      select(<<-SQL)
```
Comment (lowjoel) [inline app/models/course/group.rb]: @weiqingtoh does:
```ruby
select { cast(coalesce(..).as float) / greatest(...) }
```
work?

File: app/models/course/group.rb [hunk 4]
```diff
@@ -15,6 +15,57 @@ class Course::Group < ActiveRecord::Base
...
+  # @!attribute [r] average_achievement_count
+  #   Returns the average number of achievements obtained by group users in this group who are
+  #   students.
+  calculated :average_achievement_count, (lambda do
+    Course::GroupUser.where { group_id == course_groups.id }.
+      joins { course_user.course_user_achievements.outer }.
+      where { course_user.role == CourseUser.roles[:student] }.
+      # CAST is used to force a float division (integer division by default).
+      # greatest(#, 1) is used to avoid division by 0.
+      select(<<-SQL)
```
Comment (lowjoel) [inline app/models/course/group.rb]: same here.

File: spec/models/course/group_spec.rb [hunk 5]
```diff
@@ -93,5 +95,118 @@
...
+    end
+
+    describe '#average_experience_points' do
+      subject { group.average_experience_points }
+
+      context 'when there are no group users' do
+        it { is_expected.to eq(0) }
+      end
+
+      context 'when there is 1 or more group users' do
```
Comment (lowjoel) [inline spec/models/course/group_spec.rb]: ...are one or more... lol :stuck_out_tongue:

**Comment by lowjoel** [comment]
@weiqingtoh I don't see any route changes?

**Comment by weiqingtoh** [comment]
@lowjoel removed that commit in the end as that was handled in the other group_leaderboard PR. Was finding a nicer named URL path helper, but settled with `group_course_leaderboard_path` in the end. Sorry should have changed the title.

**Comment by lowjoel** [comment]
There usually is a way to make the route nicer, it's just a tedious process. For Rails, new actions are typically named as a verb (like get/post/index/delete), so that's why it's prefixed.

**Comment by weiqingtoh** [comment]
What I intended was to have:
```
...
  resource :leaderboard, only: [:show] 
  get 'groups' => 'leaderboard#groups', as: :group_leaderboard
...
```

That would make the named route neater => 'course_group_leaderboard', rather than the current `group_course_leaderboard`. But it doesn't get nested into the leaderboard resource, so I just did the workaround first >_<

File: app/models/course/group.rb (hunk 3 above)
Comment (weiqingtoh) [inline app/models/course/group.rb]: Tried it, doesn't work though. The issue is the syntax of `cast` - only methods I saw dealing with casting in rails >_<

Comment (lowjoel) [inline app/models/course/group.rb]: Maybe file a bug with the squeel gem, then.

Comment (weiqingtoh) [inline app/models/course/group.rb]: Oh wait. hold on let me try again. I didn't see the `.as float`

Comment (weiqingtoh) [inline app/models/course/group.rb]: It works! Will do a consolidated PR to fix all the small stuff!

**Comment by lowjoel** [comment]
It's usually some nasty combination of resource and module... I usually get it by trial and error.

**Comment by weiqingtoh** [comment]
When I get some inspiration, I'll try that!
